angr/engines/pcode/lifter.py

```diff
@@ -0,0 +1,1506 @@
+# FIXME:
+#     - Eliminate Vex references where possible
+#     - Consider moving pieces of lifter classes to higher abstraction layer
+#       to reduce duplication with Vex
+#     - Fix default_exit_target
+#     - Fix/remove NotImplementedError's
+
+from collections import defaultdict
+import pypcode
+import cle
+from archinfo import ArchARM
+from cachetools import LRUCache
+import logging
+import os
+import sys
+
+from.behavior import BehaviorFactory
+from ..engine import SimEngineBase
+from ...state_plugins.inspect import BP_AFTER, BP_BEFORE
+from ...misc.ux import once
+from ...errors import SimEngineError, SimTranslationError, SimError
+from ... import sim_options as o
+
+l = logging.getLogger(__name__)
+
+IRSB_MAX_SIZE = 400
+IRSB_MAX_INST = 99
+MAX_INSTRUCTIONS = 99999
+MAX_BYTES = 5000
+
+
+class PyVEXError(Exception):
+    pass
+
+
+class SkipStatementsError(PyVEXError):
+    pass
+
+
+class LiftingException(Exception):
+    pass
+
+
+class NeedStatementsNotification(LiftingException):
+    """
+    A post-processor may raise a NeedStatementsNotification if it needs to work with statements, but the current IRSB
+    is generated without any statement available (skip_stmts=True). The lifter will re-lift the current block with
+    skip_stmts=False upon catching a NeedStatementsNotification, and re-run the post-processors.
+
+    It's worth noting that if a post-processor always raises this notification for every basic block without statements,
+    it will essentially disable the skipping statement optimization, and it is bad for performance (especially for
+    CFGFast, which heavily relies on this optimization). Post-processor authors are encouraged to at least filter the
+    IRSBs based on available properties (jumpkind, next, etc.). If a post-processor must work with statements for the
+    majority of IRSBs, the author should implement it in PyVEX in C for the sake of a better performance.
+    """
+
+
+class IRSB:
```

Comment:
I'm not sure how I'm feeling towards having a new IRSB class whose code is mostly copied from pyvex.

First of all, @rhelmot should this file live in angr or in pyvex?

Second, is there any way to have *one* `IRSB` class across angr and pyvex instead of having two IRSB classes that are used interchangeably inside angr? This name collision (`pcode.lifter.IRSB` and `pyvex.IRSB`) really bothers me.

Third, regardless of the final decision, let's at least reuse the error classes from pyvex.

I think this is the right time to refactor pyvex a bit so that we can have a unified IRSB base class that supports both VEX IRSB and P-code IRSB.

Reply:
> I'm not sure how I'm feeling towards having a new IRSB class whose code is mostly copied from pyvex. ... is there any way to have one IRSB class across angr and pyvex instead of having two IRSB classes that are used interchangeably inside angr? This name collision (pcode.lifter.IRSB and pyvex.IRSB) really bothers me. I think this is the right time to refactor pyvex a bit so that we can have a unified IRSB base class that supports both VEX IRSB and P-code IRSB.

I agree that a unified IRSB base class should be the target. I've opted to implement it this way, with a compatible IRSB, to minimize disruptions in this PR. My intention here is to continue with the way I have implemented, and then refactor both engines to use a unified base in a follow up PR.

There is surely a way to have one IRSB base class across angr, but this will involve refactoring significant amounts of pyvex to support this. My preference is to defer this to a follow-up PR which refactors both P-code and VEX engines to use a unified IRSB base class. If you absolutely want me to do this refactor now before merge, I can do so--but it will cause further delay and expand the scope of this PR, which is already quite large.

> let's at least reuse the error classes from pyvex

I can do this for now, but in the future it would be nice to refactor those also into a lifter-agnostic error message.

Reply:
> I can do this for now, but in the future it would be nice to refactor those also into a lifter-agnostic error message.

I agree with you that we want to have another PR specifically for the refactor of pyvex and this new engine to reduce code duplication. However, having lifter-agnostic error messages is probably not too much work to do. Do you want to make that happen as part of this PR?

The less class with duplicated names, the better!

Reply:
Now using pyvex exception classes